eip-0014.md

```diff
@@ -0,0 +1,752 @@
+    }
+
+    val validSwap =
+        if (deltaReservesX > 0)
+            reservesY0.toBigInt * deltaReservesX * feeNum0 >= -deltaReservesY * (reservesX0.toBigInt * FeeDenom + deltaReservesX * feeNum0)
+        else
+            reservesX0.toBigInt * deltaReservesY * feeNum0 >= -deltaReservesX * (reservesY0.toBigInt * FeeDenom + deltaReservesY * feeNum0)
+
+    val validAction =
+        if (deltaSupplyLP == 0)
+            validSwap
+        else
+            if (deltaReservesX > 0 && deltaReservesY > 0) validDepositing
+            else validRedemption
+
+    sigmaProp(
+        validSuccessorScript &&
+        preservedFeeConfig &&
+        preservedErgs &&
+        preservedPoolNFT &&
+        validLP &&
+        validPair &&
+        noMoreTokens &&
+        validAction
+    )
+}
+```
+
+#### Swap proxy-contract
+
+Swap contract ensures a swap is executed fairly from a user's perspective. The contract checks that:
+* Assets are swapped at actual price derived from pool reserves. `X_output = X_reserved * Y_input * fee_num / (Y_reserved * 1000 + Y_input * fee_num)`
+* Fair amount of DEX fee held in ERGs. `F = X_output * F_per_token`
+* A minimal amount of quote asset received as an output in order to prevent front-running attacks.
+
+Once published swap contracts are tracked and executed by ErgoDEX bots automatically. Until a swap is executed it can be cancelled by a user who created it by simply spending the swap UTXO.
+
+##### Contract parameters:
+Constant            | Type       | Description
+--------------------|------------|---------------
+Pk                  | ProveDLog  | User PublicKey
+FeeNum              | Long       | Pool fee numerator (must taken from pool params)
+QuoteId             | Coll[Byte] | Quote asset ID
+MinQuoteAmount      | Long       | Minimal amount of quote asset
+DexFeePerTokenNum   | Long       | Numerator of the DEX fee in nanoERGs per one unit of quote asset
+DexFeePerTokenDenom | Long       | Denominator of the DEX fee in nanoERGs per one unit of quote asset
+PoolNFT             | Coll[Byte] | ID of the pool NFT (Used as a reference to a concrete unique pool)
+
+```scala
+{
+    val FeeDenom = 1000
+
+    val poolIn = INPUTS(0)
+
+    val validTrade =
+        if (INPUTS.size == 2 && poolIn.tokens.size == 4) {
+            val base       = SELF.tokens(0)
+            val baseId     = base._1
+            val baseAmount = base._2.toBigInt
+
+            val poolNFT    = poolIn.tokens(0)._1
+            val poolAssetX = poolIn.tokens(2)
+            val poolAssetY = poolIn.tokens(3)
+
+            val validPoolIn = poolNFT == PoolNFT
+
+            val rewardBox     = OUTPUTS(1)
+            val quoteAsset    = rewardBox.tokens(0)
+            val quoteAmount   = quoteAsset._2.toBigInt
+            val fairDexFee    = rewardBox.value >= SELF.value - quoteAmount * DexFeePerTokenNum / DexFeePerTokenDenom
+            val relaxedOutput = quoteAmount + 1L // handle rounding loss
```

Comment:
We can imagine the case when there are no much tokens in the pool at all (maybe not so many issued), then this relaxation becomes significant?